services/httpd/msgp_emitter.go

```diff
@@ -0,0 +1,160 @@
+package httpd
+
+import (
+	"io"
+	"net/http"
+
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/tinylib/msgp/msgp"
+)
+
+type messagePackEncoder struct {
+	Epoch     string
+	ChunkSize int
+}
+
+func (e *messagePackEncoder) ContentType() string {
+	return "application/x-msgpack"
+}
+
+func (e *messagePackEncoder) Encode(w io.Writer, header ResponseHeader, results <-chan *influxql.ResultSet) {
+	var convertToEpoch func(row *influxql.Row)
+	if e.Epoch != "" {
+		convertToEpoch = epochConverter(e.Epoch)
+	}
+	values := make([][]interface{}, 0, e.ChunkSize)
+
+	enc := msgp.NewWriter(w)
+	enc.WriteMapHeader(1)
+	enc.WriteString("results")
+	enc.WriteInt(header.Results)
+
+	enc.Flush()
+	if w, ok := w.(http.Flusher); ok {
+		w.Flush()
+	}
+
+	for result := range results {
+		var messages []Message
+		if len(result.Messages) > 0 {
+			messages = make([]Message, len(result.Messages))
+			for i, m := range result.Messages {
+				messages[i].Level = m.Level
+				messages[i].Text = m.Text
+			}
+		}
+		header := ResultHeader{
+			ID:       result.ID,
+			Messages: messages,
+		}
+		if result.Err != nil {
+			err := result.Err.Error()
+			header.Error = &err
+		}
+		header.EncodeMsg(enc)
+		if result.Err != nil {
+			enc.Flush()
+			if w, ok := w.(http.Flusher); ok {
+				w.Flush()
+			}
+			continue
+		}
+
+		for series := range result.SeriesCh() {
+			enc.WriteArrayHeader(2)
+			enc.WriteInt(1)
+			enc.WriteBool(true)
+
+			if series.Err != nil {
+				err := series.Err.Error()
+				header := SeriesError{Error: err}
+				header.EncodeMsg(enc)
```

Comment:
The reason why I do this is because of a limitation with `tinylib/msgp`. The library doesn't take kindly to not encoding certain arguments and will instead encode a blank string with them instead of just omitting the entry. So I use two structures that are supposed to be one structure to trick msgp into generating the correct code.

The intention is that the `error` key is part of the `SeriesHeader`.